cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (C) 2018 Acutam Automation, LLC
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/*
+ * @ingroup cpu_atmega_common
+ * @ingroup drivers_periph_rtt
+ * @{
+ *
+ * @file
+ * @brief       Low-level ATmega RTT driver implementation
+ *
+ * @author      Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * In order to safely sleep when using the RTT:
+ * 1. Disable interrupts
+ * 2. Write to one of the asynch registers (e.g. TCCR2A)
+ * 3. Wait for ASSR register's busy flags to clear
+ * 4. Re-enable interrupts
+ * 5. Sleep before interrupt re-enable takes effect
+ *
+ * @}
+ */
+
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "periph/rtt.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+typedef struct {
+    uint16_t ext_comp;          /* Extend compare to 24-bits */
+    rtt_cb_t alarm_cb;          /* callback called from RTT alarm */
+    void *alarm_arg;            /* argument passed to the callback */
+    rtt_cb_t overflow_cb;       /* callback called when RTT overflows */
+    void *overflow_arg;         /* argument passed to the callback */
+} rtt_state_t;
+
+static rtt_state_t rtt_state;
+
+/* Extend TCNT2 to 24-bits. This is treated as-if it is a register as it is
+ * incremented by the overflow ISR. This MUST BE marked volatile for two
+ * reasons: to guarantee ordering with other register operations, and to
+ * force redundant access to occur (in case it was changed by the ISR). */
+static volatile uint16_t ext_cnt;
```

Comment:
I reckon you suggest replacing `volatile uint16_t ext_cnt` with `atomic_uint_fast16_t`.
Do you know how this is implemented? Will it just disable interrupts for every access to the variable?

Reply:
> Will it just disable interrupts for every access to the variable?

On AVR for accesses to > 8bit variables: Yes.

(On systems that would allow to access an 16 bit wide type with a single atomic machine instruction, IRQs would not be disabled. And if needed, memory barrier machine instructions would be inserted. So C11 atomics are either faster, or about as fast as `unsigned state = irq_disable(); foo = 42; irq_restore(state)`. However, when writing to (or reading from) more than one variable in a row, `irq_disable()` and `irq_restore() are better: The whole transaction enclosed by them would be atomic and only one IRQ-disable and one IRQ-restore instruction are needed.)

Reply:
I think the `irq_disable()` option is better here. In the access function to the RTT from thread context one then can be sure that the uper 16 bits of the virtual 24 bit counter have not changed during read and that a potential IRQ flag for the timer overflow is in regard to that and the lower 8 bits just read. If after accessing the lower 8 bits (the hardware register) the IRQ flag for the overflow is set, one just can increment the local copy of the upper 16 bits by one and set the local copy of the lower 8 bits to zero, if it was 255. No need for a loop anymore :-)

The chance that the lower 8 bits have been increment by hardware two ticks during that time is zero, as the clock ticks rather slow and now other threads or ISRs can be served in between because of the disabled interrupts. It could happen that the read of the lower 8 bits has yielded 255 and the overflow happend just when reading the overflow interrupt flag. In that case the local copy of the upper bits have to be incremented and the local copy of the lower bits set to zero. If the IRQ flag was set and the lower bits were e.g. 3, the IRQ flag was set for some time and the RTT ISR was not hanled due to load for ages. This could happen when  the current RTT value is read from a high prio and slow ISR that prevented the overflow RTT ISR to be served for some time. (A stupid idea to have slow ISRs, but as this is trivial to to handle, I'd say the RTT implementation can just handle this.)